docs: clarify progressive rollout gate scenarios (#78314)
airbytehq/airbyte: .github/progressive-rollout-gate-comment.md

</details>

<details><summary>If the connector version increments to a higher `-rc` version...</summary>

After this PR is merged, the new RC will be published and registered, replacing the active RC marker. When the new RC is registered, the platform cancels any existing non-terminal rollout for this connector without unpinning actors.

After merging, you still need to start the new rollout. During start, pinned actors from the previous rollout can be moved to the new RC.

</details>

<details><summary>If the connector version changes from RC to non-RC (GA) version...</summary>

> This PR should not be merged while the RC rollout is active. First finalize the active rollout as successful or cancel it.

Author: ```suggestion
You should not merge the PR unless/until the RC has been finalized as canceled. See above `Rollout state` for detected status.

> [!Warning]
> This PR should not be merged if the RC rollout is still active. First finalize the active rollout as successful or cancel it.
```

> If the rollout is finalized as successful, the platform triggers a promotion workflow that strips the `-rc` suffix, removes stable-version `registryOverrides`, disables progressive rollout, merges that promotion, and unpins actors.

Author: ```suggestion
When you finalize an RC rollout as successful, the platform triggers a promotion workflow that strips the `-rc` suffix, removes stable-version `registryOverrides`, disables progressive rollout, force-merges that promotion, and unpins actors.
```

After merging, you still need to start the new rollout. During start, pinned actors from the previous rollout can be moved to the new RC.

</details>

<details><summary>If the connector version changes from RC to non-RC (GA) version...</summary>

This PR should not be merged while the RC rollout is active. First finalize the active rollout as successful or cancel it.

- If the rollout is finalized as successful, the platform triggers a promotion workflow that strips the `-rc` suffix, removes stable-version `registryOverrides`, disables progressive rollout, merges that promotion, and unpins actors.
- If this PR is merged anyway, it should remove stable-version `registryOverrides` and disable progressive rollout so the published GA version becomes the default for eligible non-pinned actors.
- The active rollout is not stopped immediately by the merge, and actors already pinned by the rollout remain pinned until the rollout later finalizes or is canceled.

> - If the rollout later finalizes as promoted and the platform default no longer matches the rollout's RC version stem, the rollout is canceled as superseded.

Author: Delete these bullets. No longer relevant since we're discouraging merge in that condition.

This PR should not be merged while the RC rollout is active. First finalize the active rollout as successful or cancel it.

- If the rollout is finalized as successful, the platform triggers a promotion workflow that strips the `-rc` suffix, removes stable-version `registryOverrides`, disables progressive rollout, merges that promotion, and unpins actors.
- If this PR is merged instead, the active rollout is not stopped immediately and actors already pinned by the rollout remain pinned until the rollout later finalizes or is canceled.

</details>

<details><summary>If the connector version changes from RC to a new (distinct) GA version...</summary>

This PR should not be merged while the RC rollout is active. First finalize or cancel the active rollout.

- If this PR is merged anyway, it should remove stable-version `registryOverrides` and disable progressive rollout so the published GA version becomes the default for eligible non-pinned actors.
- The active rollout is not stopped immediately by the merge, and actors already pinned by the rollout remain pinned until the rollout later finalizes or is canceled.

> - If the rollout later finalizes as promoted and the platform default is still this distinct GA version, the rollout is canceled as superseded.

Author: Re-join these forks. No remaining reason to keep separate.

</details>

> <details><summary>If the connector version increments to a higher `-rc` version...</summary>

Author: Reorder this to be the 2nd fork (now of 3), moving up one position.

</details>

<details><summary>If the connector version changes to the active RC's version stem...</summary>

The active rollout is not stopped immediately by the merge.

- If this PR removes the stable-version `registryOverrides` and disables progressive rollout, the published GA version can become the default for eligible non-pinned actors.

> - If this PR leaves stable-version `registryOverrides` in place or leaves progressive rollout enabled, non-rollout registry consumers should continue to see the pre-RC stable version.

Author: What's the correct disposition for `registryOverrides`? Let's be prescriptive and then simplify the condition.

- Actors already pinned by the active rollout remain pinned; publishing the GA version does not immediately unpin them.
- If the rollout later finalizes as promoted and the platform default matches this version, the rollout succeeds and unpins actors.

</details>

<details><summary>If the connector version changes from RC to a new (distinct) GA version...</summary>

The active rollout is not stopped immediately by the merge.

- If this PR removes the stable-version `registryOverrides` and disables progressive rollout, the published GA version can become the default for eligible non-pinned actors.

> - If this PR leaves stable-version `registryOverrides` in place or leaves progressive rollout enabled, non-rollout registry consumers should continue to see the pre-RC stable version.

Author: Ditto.

<details><summary>If the connector version changes from RC to non-RC (GA) version...</summary>

The merged PR may publish a GA version, but the active rollout is not stopped immediately by the merge.

During an active RC rollout, `registryOverrides.cloud.dockerImageTag` and `registryOverrides.oss.dockerImageTag` usually pin non-rollout registry consumers to the pre-RC stable version. Older registry generation depended on those overrides to prevent the RC from becoming default. The newer `airbyte-ops registry compile` path is semver-aware, so `-rc` versions are not eligible as GA/default/latest versions. It also skips GA versions whose metadata still has progressive rollout enabled. If this PR removes those overrides and disables progressive rollout, the newly published GA version can become the registry default for eligible non-pinned actors.

> If the active rollout is later promoted, the platform verifies that the default version matches this rollout's RC stem. For example, `1.2.3-rc.1` is expected to become `1.2.3`. If the default is a different GA version, the rollout is marked `CANCELED` as superseded. If the default matches the RC stem, the rollout can complete as promoted.

Author: This type of documentation might be appropriate for hte detailed reference documentation. It's not appropriate here. We need to concicely answer "what happens if I merge it?" and have effective branching that is easy to read.

I propose we may want to inject conditions based on if `registryOverrides` in this PR contains version overrides or not.

</details>

<details><summary>If the connector version increments to a higher `-rc` version...</summary>

> After this PR is merged, it may publish a new release candidate and replace the active RC marker. When the new RC is registered, the platform cancels any existing non-terminal rollout for this connector. After that, when the new rollout is started through the normal rollout flow, actors pinned by the previous rollout are moved to the new RC.

Author: Is this accurate?

```suggestion
After this PR is merged, the new RC will replace the active RC and the active RC will ba canceled (if still active).

After merging, you will still need to activate the new RC version. During activation, you have the option to migrate pinned actors, moving them automatically to the new RC.
```

Author: > Updated along these lines. I kept “may be published and registered” because the gate comment is describing the result of merging the PR plus the downstream publish/registry steps, not something guaranteed by GitHub merge alone.

This is idiotic. Yes, we understand the universe has entropy. That doesn't prevent us from setting clear expectations. Unless there are known reasons why it shouldn't follow the expected processes (besides random failure) then we ned to set clear expectations. I hedged when I didn't know the process, but now is the time to get clarity and not hedge.

<details><summary>If the connector version changes from RC to non-RC (GA) version...</summary>

The active rollout is not stopped immediately by the merge.

- If this PR removes the stable-version `registryOverrides` and disables progressive rollout, the published GA version can become the default for eligible non-pinned actors.
- If this PR leaves stable-version `registryOverrides` in place or leaves progressive rollout enabled, non-rollout registry consumers should continue to see the pre-RC stable version.

> - Actors already pinned by the active rollout remain pinned until that rollout finalizes or is canceled.

Author: Should we split this condition to fork on whether the GA is the RC's stem? E.g. if RC is 1.2.3-rc1, and if we merge 1.2.3, then I think in that case it does auto-cancel or auto-promote (you tell me which) the RC.

<details><summary>If the connector version changes from RC to non-RC (GA) version...</summary>

The active rollout is not stopped immediately by the merge.

- If this PR removes the stable-version `registryOverrides` and disables progressive rollout, the published GA version can become the default for eligible non-pinned actors.
- If this PR leaves stable-version `registryOverrides` in place or leaves progressive rollout enabled, non-rollout registry consumers should continue to see the pre-RC stable version.
- Actors already pinned by the active rollout remain pinned; publishing the GA version does not immediately unpin them, even when the GA version matches the RC stem.

> - If the rollout later finalizes as promoted and the platform default matches the RC stem, the rollout succeeds and unpins actors. If the platform default does not match the RC stem, the rollout is canceled as superseded.

Author: Promote this as its own details/summary section. Outcome is completely different and we need to call it out. So that would be the 2nd case and this block would become the fourth. Use these headers:

<details><summary>If the connector version changes to the active RC's version stem...</summary>
...
</details>

<details><summary>If the connector version changes from RC to a new (distinct) GA version...</summary>
...
</details>
